synapse/storage/databases/state/store.py

```diff
@@ -404,6 +406,124 @@ def _insert_into_cache(
+                    "Trying to persist state with unpersisted prev_group: %r"
+                    % (prev_group,)
+                )
+
+            num_state_groups = len(events_and_context)
+
+            state_groups = self._state_group_seq_gen.get_next_mult_txn(
+                txn, num_state_groups
+            )
+
+            for index, (event, context) in enumerate(events_and_context):
+                context.state_group_after_event = state_groups[index]
+                # The first prev_group will be the last persisted state group, which is passed in
+                # else it will be the group most recently assigned
+                if index > 0:
+                    context.prev_group_for_state_group_after_event = state_groups[
+                        index - 1
+                    ]
+                    context.state_group_before_event = state_groups[index - 1]
+                else:
+                    context.prev_group_for_state_group_after_event = prev_group
+                    context.state_group_before_event = prev_group
+                context.delta_ids_to_state_group_after_event = {
+                    (event.type, event.state_key): event.event_id
+                }
+                context.state_delta_due_to_event = {
+                    (event.type, event.state_key): event.event_id
+                }
+                index += 1
+
+            self.db_pool.simple_insert_many_txn(
+                txn,
+                table="state_groups",
+                keys=("id", "room_id", "event_id"),
+                values=[
+                    (context.state_group_after_event, room_id, event.event_id)
+                    for event, context in events_and_context
+                ],
+            )
+
+            self.db_pool.simple_insert_many_txn(
+                txn,
+                table="state_group_edges",
+                keys=("state_group", "prev_state_group"),
+                values=[
+                    (
+                        context.state_group_after_event,
+                        context.prev_group_for_state_group_after_event,
+                    )
+                    for _, context in events_and_context
+                ],
+            )
+
+            for _, context in events_and_context:
+                assert context.delta_ids_to_state_group_after_event is not None
+                self.db_pool.simple_insert_many_txn(
+                    txn,
+                    table="state_groups_state",
+                    keys=("state_group", "room_id", "type", "state_key", "event_id"),
+                    values=[
+                        (
+                            context.state_group_after_event,
+                            room_id,
+                            key[0],
+                            key[1],
+                            state_id,
+                        )
+                        for key, state_id in context.delta_ids_to_state_group_after_event.items()
+                    ],
```

Comment:
Can we move `for _, context in events_and_context` into this so that we only call `simple_insert_many_txn` once?